### ISSUE TITLE
Compile times with large sums

### ISSUE THREAD

**Issue by andrewthad** [goldfirere/singletons#165]
I have a data type with around 80 data constructors and derived instances for Eq,Ord,Bounded, and Enum. With ghc-7.10 and singletons-2.1, this compiled in somewhere near 15 seconds. With GHC8 and singletons-2.2, it takes closer to 100 seconds. I'm not sure if this is more likely to be a singletons issue or a more general issue with GHC compile times, but I figured I would bring it up anyway. I've created a gist with a minimal example: https://gist.github.com/andrewthad/b3cb6aca69000ac342d4eee24c4e6eb1

Also, I'm a little surprised that it's able to compile without `TypeInType` being enabled explicitly, but I guess none of the generated code needs it.

**Comment by jstolarek** [comment]
Yes, this looks like GHC's fault. See [#8095](https://ghc.haskell.org/trac/ghc/ticket/8095). There are many related tickets, e.g. [#12506](https://ghc.haskell.org/trac/ghc/ticket/12506), [#11598](https://ghc.haskell.org/trac/ghc/ticket/11598)

**Comment by goldfirere** [comment]
I just tested GHC 8 with singletons-2.1, and it's slow. So I'm with @jstolarek on this one. Will keep ticket open, though, so that when GHC fixes its end, we can test.
